src/utils/reporting_workflow.js

```diff
@@ -45,13 +50,130 @@ const IG_REGISTRY = {
+/**
+ * Retrieve data from the EHR
+ *
+ * @param {string} uri - the search query for the data desired
+ * @returns axios promise with data
+ */
+async function readFromEHR(uri) {
+  const url = process.env.EHR;
+  const token = await connectToServer(url);
+  const headers = { Authorization: `Bearer ${token}` };
+  return axios
+    .get(`${url}/${uri}`, { headers: headers })
+    .then(response => response.data)
+    .catch(err => console.error(err));
+}
+
+/**
+ * Post the reporting bundle to the endpoint designated in the PlanDefinition
+ *
+ * @param {PlanDefinition} planDefinition - the current PlanDefinition from the context
+ * @param {Bundle} bundle - the reporting Bundle to submit
+ * @returns axios promise with data
+ */
+async function submitBundle(planDefinition, bundle) {
+  const receiverAddress = planDefinition.extension.find(e => e.url === RECEIVER_ADDRESS_EXT);
+  const endpointRef = receiverAddress.valueReference.reference;
+  const endpointId = endpointRef.split('/')[1];
+  const endpoint = db.select('endpoints', e => e.id === endpointId);
+  const url = endpoint.address;
+  // Not sure how this will work since the url comes from endpoint.address....
+  const token = await connectToServer(url);
+  const headers = { Authorization: `Bearer ${token}` };
+  return axios
+    .post(url, bundle, { headers: headers })
+    .then(response => response.data)
+    .catch(err => console.error(err));
+}
+
+/**
+ * Helper function to call specific operations on the data trust service server.
+ *
+ * @param {string} operation = 'deidentify' | 'anonymize' | 'pseudonymize'
+ * @param {Bundle} bundle - the bundle to perform operation on
+ * @returns axios promise with data
+ */
+function dataTrustOperation(operation, bundle) {
+  return axios
+    .post(`${process.env.DATA_TRUST_SERVICE}/Bundle/$${operation}`, bundle)
+    .then(response => response.data)
+    .catch(err => console.error(err));
+}
+
+/**
+ * Function to use the PlanDefinition and triggering resource to create a report
+ *
+ * @param {PlanDefinition} planDef - the PlanDefinition resource
+ * @param {*} resource - the resource which triggered the notification
+ */
+async function initializeReportingWorkflow(planDef, resource = null) {
+  // TODO: MEDMORPH-49 to make sure the resource is always included
+  if (!resource) return;
+
+  let patient = null;
+  if (resource.patient) patient = await getReferencedResource(resource.patient.reference);
+  else if (resource.subject) patient = await getReferencedResource(resource.subject.reference);
+  else if (resource.resourceType === 'Patient') patient = resource;
+
+  let encounter = null;
+  if (resource.encounter) encounter = await getReferencedResource(resource.encounter.reference);
+  else if (resource.context) encounter = await getReferencedResource(resource.context.reference);
+  else if (resource.resourceType === 'Encounter') encounter = resource;
+
+  // QUESTION: Should encounter and patient be saved to the database?
```

Comment:
That's a good question. Since the app in practice would be working with PHI/PII there may be some considerations about storing that.
I also realize that we're storing functions (for the various clients) in the context, which isn't going to work if we want to reload the context. Now I'm thinking maybe we don't want to store the context in the DB at all.

Reply:
That's a good point. I can refactor the client functions so they are not included in the context and then update the actions to use those functions. For example, the client.source can just be the endpoint and the submit function defined in action.js

Reply:
Refactored the client methods but haven't done anything regarding storing the encounter or patient.